### PR TITLE
Switched to use dynamic shared informer for Garbage Collector.

### PR DESCRIPTION
**Pull request by yliaog**
**What type of PR is this?**
> Uncomment only one ` /kind <>` line, hit enter to put that in a new line, and remove leading whitespaces from that line:
>
> /kind api-change
> /kind bug
/kind cleanup
> /kind design
> /kind documentation
> /kind failing-test
> /kind feature
> /kind flake

**What this PR does / why we need it**:
Switched to use dynamic shared informer for Garbage Collector.

**Which issue(s) this PR fixes**:

prerequisite for #53777
**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

### CODE REVIEW

File: pkg/controller/garbagecollector/garbagecollector_test.go [hunk 1]
```diff
@@ -430,34 +435,6 @@ func TestDependentsRace(t *testing.T) {
 }
 
 // test the list and watch functions correctly converts the ListOptions
```
Comment (caesarxuchao) [inline pkg/controller/garbagecollector/garbagecollector_test.go]: Remove the comment as well?

**Review by caesarxuchao — COMMENTED** [review]
lgtm except for a nit.

File: pkg/controller/garbagecollector/garbagecollector_test.go (hunk 1 above)
Comment (yliaog) [reply pkg/controller/garbagecollector/garbagecollector_test.go]: removed

File: pkg/controller/garbagecollector/garbagecollector_test.go [hunk 2]
```diff
@@ -60,6 +62,7 @@ func TestGarbageCollectorConstruction(t *testing.T) {
 	tweakableRM := meta.NewDefaultRESTMapper(nil)
 	rm := &testRESTMapper{meta.MultiRESTMapper{tweakableRM, testrestmapper.TestOnlyStaticRESTMapper(legacyscheme.Scheme)}}
 	dynamicClient, err := dynamic.NewForConfig(config)
+
```
Comment (caesarxuchao) [inline pkg/controller/garbagecollector/garbagecollector_test.go]: nit: remove the extra new line.

**Review by caesarxuchao — COMMENTED** [review]
/lgtm
/approve

**Comment by yliaog** [comment]
@caesarxuchao done some cleanups,  PTAL

**Comment by yliaog** [comment]
@liggitt @deads2k could you please review and if ok approve? Thanks

**Comment by deads2k** [comment]
I really like how well this fell out.  I need to spend a few minutes tomorrow double checking tests.  @yliaog you could save me a few minutes by linking me to the test checking GC of a CRD.

**Comment by yliaog** [comment]
the three test cases below in the integration test are for CRD. Thanks.

TestCustomResourceCascadingDeletion:
https://github.com/kubernetes/kubernetes/blob/9606fa5d87d58a5a1387336c67639d5825a64640/test/integration/garbagecollector/garbage_collector_test.go#L806

TestMixedRelationships:
https://github.com/kubernetes/kubernetes/blob/9606fa5d87d58a5a1387336c67639d5825a64640/test/integration/garbagecollector/garbage_collector_test.go#L866

TestCRDDeletionCascading:
https://github.com/kubernetes/kubernetes/blob/9606fa5d87d58a5a1387336c67639d5825a64640/test/integration/garbagecollector/garbage_collector_test.go#L964

**Comment by deads2k** [comment]
I'd like to see a test that does a delete of the CRD (not the CR) and a recreate of the same CRD to be sure the logic works and correctly builds out the graph.

/approve
/hold

holding for the test.  Feel free to release the hold once the test is present, I don't need a second look.

**Comment by caesarxuchao** [comment]
> I'd like to see a test that does a delete of the CRD (not the CR) and a recreate of the same CRD to be sure the logic works and correctly builds out the graph.

The test (https://github.com/kubernetes/kubernetes/pull/74440/files#diff-71e62caa4db77f62b62feb4951fd7ca8R976) was added. @yliaog could you note down the issue the test discovered, and how it was fixed by https://github.com/kubernetes/kubernetes/pull/78029?

/hold remove
/lgtm

**Comment by yliaog** [comment]
/hold

wait for https://github.com/kubernetes/kubernetes/pull/78029 to merge

**Comment by liggitt** [comment]
TestStressingCascadingDeletion started flaking as soon as this PR was merged. See #78525

**Pull request merged**